src/include/storage/store/var_list_column.h

```diff
@@ -1,43 +1,45 @@
 #pragma once
 
 #include "column.h"
+#include "var_list_column_chunk.h"
 
-// List is a nested data type which is stored as two chunks:
+// List is a nested data type which is stored as three chunks:
 // 1. Offset column (type: INT64). Using offset to partition the data column into multiple lists.
-// 2. Data column. Stores the actual data of the list.
+// 2. Size column. Stores the size of each list.
+// 3. Data column. Stores the actual data of the list.
 // Similar to other data types, nulls are stored in the null column.
 // Example layout for list of INT64:
 // Four lists: [4,7,8,12], null, [2, 3], []
 // Offset column: [4, 4, 6, 6]
+// Size column: [4, 0, 2, 0]
 // data column: [4, 7, 8, 12, 2, 3]
-// When reading the data, we firstly read the offset column and utilize the offset column to find
-// the data column partitions.
-// 1st list(offset 4): Since it is the first element, the start offset is a constant 0, and the end
-// offset is 4. Its data is stored at position 0-4 in the data column.
-// 2nd list(offset 4): By reading the null column, we know that the 2nd list is null. So we don't
-// need to read from the data column.
-// 3rd list(offset 6): The start offset is 4(by looking up the offset for the 2nd list), and the end
-// offset is 6. Its data is stored at position 4-6 in the data column.
-// 4th list(offset 6): The start offset is 6(by looking up the offset for the 3rd list), and the end
-// offset is 6. Its data is stored at position 6-6 in the data column (empty list).
+// When updating the data, we first append the data to the data column, and then update the offset
+// and size. For example, we want to update [4,7,8,12] to [1,3,4]. We first append [1,3,4] to the
+// data column, and then update the offset and size. The layout after the update: Offset column: [9,
+// 4, 6, 6] Size column: [3, 0, 3, 0] data column: [4, 7, 8, 12, 2, 3, 1, 3, 4] This design is good
+// for writing performance. But it is bad for scan performance since it may cause random access to
+// the data column. Worse, the data column will be increased every update/insert. To balance the
+// write performance and read performance, we rewrite the whole var list column in ascending order
+// when the size of data column is larger than a threshold(Currently we use capacity/2).
```

Comment:
// When updating the data, we first append the data to the data column, and then update the offset
// and size accordingly. Besides offset column, we introduce an extra size column here to enable in-place updates of a list column.
// In a list column chunk, offsets of lists are not always sorted after updates.
// This is good for writes, but it introduces extra overheads for scans, as lists can be scattered, and scans have to be broken into multiple small reads. To achieve a balance between reads and writes, during updates, we rewrite the whole var list column chunk in ascending order
// when the offsets are not sorted in ascending order and the size of data column chunk is larger than half of its capacity.